universalModule/software/RoFIDriver/src/rofi_hal.hpp

```diff
@@ -0,0 +1,344 @@
+            float maxTorque;
+        };
+
+        virtual const Capabilities& getCapabilities() = 0;
+        virtual float getVelocity() = 0;
+        virtual void setVelocity( float velocity ) = 0;
+        virtual float getPosition() = 0;
+        virtual void setPosition( float pos, float velocity,
+            std::function< void( Joint ) > callback ) = 0;
+        virtual float getTorque() = 0;
+        virtual void setTorque( float torque ) = 0;
+        virtual void onError(
+            std::function< void( Joint, Error, const std::string& ) > callback ) = 0;
+    };
+
+    /**
+     * \brief Get maximal joint poisition in radians
+     */
+    float maxPosition() const { return _impl->getCapabilities().maxPosition; }
+
+    /**
+     * \brief Get minimal joint poisition in radians
+     */
+    float minPosition() const { return _impl->getCapabilities().minPosition; }
+
+    /**
+     * \brief Get maximal joint speed in rad / s
+     */
+    float maxSpeed() const { return _impl->getCapabilities().maxSpeed; }
+
+    /**
+     * \brief Get minimal joint speed in rad / s
+     */
+    float minSpeed() const { return _impl->getCapabilities().minSpeed; }
+
+    /**
+     * \brief Get maximal joint torque in N * m
+     */
+    float maxTorque() const { return _impl->getCapabilities().maxTorque; }
+
+    /**
+     * \brief Get current joint velocity.
+     *
+     * \return velocity setpoint in rad / s
+     */
+    float getVelocity() { return _impl->getVelocity(); }
+
+    /**
+     * \brief Change the joint's control mode to velocity and start moving at
+     * velocity.
+     */
+    void setVelocity( float velocity ) { _impl->setVelocity( velocity ); }
+
+    /**
+     * \brief Get current joint position in rad
+     *
+     * This works in all three control modes - position, velocity and torque
+     * mode.
+     */
+    float getPosition() { return _impl->getPosition(); }
+
+    /**
+     * \brief Change the joint's control mode to position and move to position.
+     *
+     * \param pos required position in rad
+     * \param velocity velocity limit in rad/s, required to be positive and
+     * non-zero
+     * \param callback callback to be called once the position is reached
+     */
+    void setPosition( float pos, float velocity,
+        std::function< void( Joint ) > callback )
+    {
+        _impl->setPosition( pos, velocity, callback );
+    }
+
+    /**
+     * \brief Get the current torque of the joint.
+     *
+     * \return torque in N * m
+     */
+    float getTorque() { return _impl->getTorque(); }
+
+    /**
+     * \brief Change the joint's control mode to torque and provide given torque
+     *
+     * \param torque torque setpoint in N * m
+     */
+    void setTorque( float torque ) { return _impl->setTorque( torque ); }
+
+    /**
+     * \brief Set error handling callback
+     */
+    void onError( std::function< void( Joint, Error, const std::string& ) > callback ) {
```

Comment:
As we discussed - call all callbacks given (and add this information to the docs)